### PR TITLE
Feature: extend functional keys handling with key modifiers

### PR DESCRIPTION
**Pull request by penguinolog**
* F1 is a special and use number `1` postfixed by `P`
* F2 is a special and use number `1` postfixed by `Q`
* F3 is a special and use number `1` postfixed by `R`
* F4 is a special and use number `1` postfixed by `S`

Related #702
Related #702

##### Checklist
- [x] I've ensured that similar functionality has not already been implemented
- [X] I've ensured that similar functionality has not earlier been proposed and declined
- [X] I've branched off the `master` or `python-dual-support` branch
- [X] I've merged fresh upstream into my branch recently
- [X] I've ran `tox` successfully in local environment
- [ ] I've included docstrings and/or documentation and/or examples for my code (if this is a new feature)

### CODE REVIEW

**Comment by penguinolog** [comment]
Before:
image

After:
image

File: urwid/escape.py [hunk 1]
```diff
@@ -476,14 +494,14 @@ def process_keyqueue(codes: Sequence[int], more_available: bool) -> tuple[list[s
         else:
             return [f"<{code:d}>"], codes[1:]
 
-        for i in range(need_more):
-            if len(codes) - 1 <= i:
+        for i in range(1, need_more + 1):
```
Comment (penguinolog) [inline urwid/escape.py:497]: simplify indexing

File: urwid/escape.py [hunk 2]
```diff
@@ -459,13 +468,22 @@ def process_keyqueue(codes: Sequence[int], more_available: bool) -> tuple[list[s
 
     em = str_util.get_byte_encoding()
 
-    if em == "wide" and code < 256 and within_double_byte(code.to_bytes(1, "little"), 0, 0):
+    if (
```
Comment (penguinolog) [inline urwid/escape.py:471]: force fit in 12 symbols

File: urwid/escape.py [hunk 3]
```diff
@@ -173,6 +173,15 @@ def escape_modifier(digit):
             ),
         )
     ),
+    *(
+        # Modified F1, F5-F12 - [XX;#P form
+        (f"[{num};{digit}P", escape_modifier(digit) + f"f{num}")
+        for digit in "12345678"
+        for num in (1, range(5, 12))
+    ),
+    *((f"[1;{digit}Q", escape_modifier(digit) + "f2") for digit in "12345678"),  # F2 is special
```
Comment (penguinolog) [inline urwid/escape.py:182]: incorrect parse of this sequence always cause exit from "input_test"

File: urwid/escape.py [hunk 4]
```diff
@@ -300,19 +324,30 @@ def read_sgrmouse_info(self, keys, more_available: bool):
...
+        x -= 1
+        y -= 1
+
+        if action == "M":
+            if b & MOUSE_DRAG_FLAG:
                 action = "drag"
             else:
                 action = "press"
-        else:
+        elif action == "m":
```
Comment (penguinolog) [inline urwid/escape.py:345]: https://invisible-island.net/xterm/ctlseqs/ctlseqs.pdf page 46

File: urwid/escape.py [hunk 5]
```diff
@@ -289,9 +312,10 @@ def read_sgrmouse_info(self, keys, more_available: bool):
                 raise MoreInputRequired()
             return None
 
-        (b, x, y) = value[:-1].split(";")
+        (b, x, y) = (int(val) for val in value[:-1].split(";"))
```
Comment (penguinolog) [inline urwid/escape.py:315]: let's cast to `int` in-place

File: urwid/escape.py [hunk 6]
```diff
@@ -154,10 +158,11 @@ def escape_modifier(digit):
         for letter, key in zip("ABCDEFGH", ("up", "down", "right", "left", "5", "end", "5", "home"))
     ),
     *(
-        # modified F1-F4 keys -- O#X form
-        (f"O{digit}{letter}", escape_modifier(digit) + key)
+        # modified F1-F4 keys - O#X form and [1;#X form
+        (prefix + digit + letter, escape_modifier(digit) + f"f{number}")
+        for prefix in ("O", "[1;")
```
Comment (penguinolog) [inline urwid/escape.py:163]: figured out, that generator for another prefix was available

**Comment by wardi** [comment]
think it's worth adding tests of the parsing code here to make sure it doesn't break?

**Comment by penguinolog** [comment]
Probably yes. will add in 1 hour and fix "ruff" checker

**Pull request merged**